pkg/kubelet/pending_and_terminating_handler.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubelet
+
+import (
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/kubelet/lifecycle"
+	"k8s.io/kubernetes/pkg/kubelet/status"
+)
+
+const (
+	deletedWhilePendingReason  = "Deleted"
+	deletedWhilePendingMessage = "Deleted while in the pending phase"
+)
+
+// pendingAndTerminatingHandler determines pending and terminating pods to be evicted
+type pendingAndTerminatingHandler struct {
+	// the provider of pod status
+	podStatusProvider status.PodStatusProvider
+	// the recorder to dispatch events when we identify a pod to be evicted
+	recorder record.EventRecorder
+}
+
+// newPendingAndTerminatingHandler
+func newPendingAndTerminatingHandler(
+	podStatusProvider status.PodStatusProvider,
+	recorder record.EventRecorder,
+) (*pendingAndTerminatingHandler, error) {
+	// check for all required fields
+	if podStatusProvider == nil {
+		return nil, fmt.Errorf("pod status provider must not be nil")
+	}
+	if recorder == nil {
+		return nil, fmt.Errorf("event recorder must not be nil")
+	}
+	return &pendingAndTerminatingHandler{
+		podStatusProvider: podStatusProvider,
+		recorder:          recorder,
+	}, nil
+}
+
+// ShouldSync returns true if the pod should be evicted.
+func (m *pendingAndTerminatingHandler) ShouldSync(pod *v1.Pod) bool {
+	return m.shouldBeEvicted(pod)
+}
+
+// ShouldEvict returns true if the pod should be evicted.
+func (m *pendingAndTerminatingHandler) ShouldEvict(pod *v1.Pod) lifecycle.ShouldEvictResponse {
+	if !m.shouldBeEvicted(pod) {
+		return lifecycle.ShouldEvictResponse{Evict: false}
+	}
+	klog.V(5).InfoS("Deleting a pending and terminating pod", "pod", klog.KObj(pod))
```

Comment:
general comment around naming: We should avoid using the "terminating" in the log and struct names. Terminating in context of pod lifecycles is referring to pods that are going through their termination phases (i.e `syncTerminatingPod`). However, these pods were never started (hence they are in Pending), so we should reduce confusion here.

Reply:
Renamed the file to `deleted_pods_handler`, and rename the associated functions.